# Предобработка датасета

## Обзор данных

In [30]:
import pandas as pd
import numpy as np
import ast

In [31]:
df = pd.read_csv('games.csv')

In [32]:
df.sample()

Unnamed: 0                    Title  Release Date        Team  Rating   
714         714  Castlevania: Bloodlines  Mar 17, 1994  ['Konami']     3.7  \

    Times Listed Number of Reviews                     Genres   
714          447               447  ['Adventure', 'Platform']  \

                                               Summary   
714  Time refuses to forget the Belmont family's ho...  \

                                               Reviews Plays Playing Backlogs   
714  ["As a child who constantly bounced his leg wh...  1.8K      25      682  \

    Wishlist  
714      230

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1512 entries, 0 to 1511
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         1512 non-null   int64  
 1   Title              1512 non-null   object 
 2   Release Date       1512 non-null   object 
 3   Team               1511 non-null   object 
 4   Rating             1499 non-null   float64
 5   Times Listed       1512 non-null   object 
 6   Number of Reviews  1512 non-null   object 
 7   Genres             1512 non-null   object 
 8   Summary            1511 non-null   object 
 9   Reviews            1512 non-null   object 
 10  Plays              1512 non-null   object 
 11  Playing            1512 non-null   object 
 12  Backlogs           1512 non-null   object 
 13  Wishlist           1512 non-null   object 
dtypes: float64(1), int64(1), object(12)
memory usage: 165.5+ KB


## Пропуски

In [34]:
df.isna().sum()

Unnamed: 0            0
Title                 0
Release Date          0
Team                  1
Rating               13
Times Listed          0
Number of Reviews     0
Genres                0
Summary               1
Reviews               0
Plays                 0
Playing               0
Backlogs              0
Wishlist              0
dtype: int64

In [35]:
df[df['Rating'].isna()==1].T

587    
Unnamed: 0                                                       587  \
Title                                              Final Fantasy XVI   
Release Date                                            Jun 22, 2023   
Team               ['Square Enix', 'Square Enix Creative Business...   
Rating                                                           NaN   
Times Listed                                                     422   
Number of Reviews                                                422   
Genres                                                       ['RPG']   
Summary            Final Fantasy XVI is an upcoming action role-p...   
Reviews                                                           []   
Plays                                                             37   
Playing                                                           10   
Backlogs                                                         732   
Wishlist                                                        2.4K   

                                       649    
Unnamed: 0                              649  \
Title                     Death Stranding 2   
Release Date                releases on TBD   
Team                 ['Kojima Productions']   
Rating                                  NaN   
Times Listed                            105   
Number of Reviews                       105   
Genres             ['Adventure', 'Shooter']   
Summary                                 NaN   
Reviews                                  []   
Plays                                     3   
Playing                                   0   
Backlogs                                209   
Wishlist                                644   

                                                                713    
Unnamed: 0                                                       713  \
Title                                      Final Fantasy VII Rebirth   
Release Date                                            Dec 31, 2023   
Team                                                 ['Square Enix']   
Rating                                                           NaN   
Times Listed                                                     192   
Number of Reviews                                                192   
Genres                                                            []   
Summary            This next standalone chapter in the FINAL FANT...   
Reviews                                                           []   
Plays                                                             20   
Playing                                                            3   
Backlogs                                                         354   
Wishlist                                                        1.1K   

                                                                719    
Unnamed: 0                                                       719  \
Title                                                      Lies of P   
Release Date                                            Aug 01, 2023   
Team                                     ['NEOWIZ', 'Round8 Studio']   
Rating                                                           NaN   
Times Listed                                                     175   
Number of Reviews                                                175   
Genres                                                       ['RPG']   
Summary            Inspired by the familiar story of Pinocchio, L...   
Reviews                                                           []   
Plays                                                              5   
Playing                                                            0   
Backlogs                                                         260   
Wishlist                                                         939   

                                                                726    
Unnamed: 0                                 

Есть игры, которые ещё только будут выходить в этом и следующем годах, поэтому их рейтинг пока неизвестен. Заменю пропуск на -1.

In [36]:
df[df['Team'].isna()==1]

Unnamed: 0                 Title  Release Date Team  Rating   
1245        1245  NEET Girl Date Night  Oct 21, 2022  NaN     2.7  \

     Times Listed Number of Reviews            Genres   
1245           21                21  ['Visual Novel']  \

                                                Summary   
1245  Your friend sets you up on a date with his NEE...  \

                                                Reviews Plays Playing   
1245  ['this sucked. "Omg she is literally me" is no...   106       1  \

     Backlogs Wishlist  
1245       44       42

Как подсказывает гугл разработчик этой игры - Hitsuji Goods. Поэтому внесём это в наши данные

In [37]:
df.loc[df['Title']=='NEET Girl Date Night','Team'] = "['Hitsuji Goods']"
df['Rating'] = df['Rating'].fillna(-1)

## Обработка даты релиза

In [38]:
df[df['Release Date'] == 'releases on TBD']

Unnamed: 0                              Title     Release Date   
644          644                          Deltarune  releases on TBD  \
649          649                  Death Stranding 2  releases on TBD   
1252        1252  Elden Ring: Shadow of the Erdtree  releases on TBD   

                                                Team  Rating Times Listed   
644                                      ['tobyfox']     4.3          313  \
649                           ['Kojima Productions']    -1.0          105   
1252  ['FromSoftware', 'Bandai Namco Entertainment']     4.8           18   

     Number of Reviews                                            Genres   
644                313  ['Adventure', 'Indie', 'Music', 'Puzzle', 'RPG']  \
649                105                          ['Adventure', 'Shooter']   
1252                18                              ['Adventure', 'RPG']   

                                                Summary   
644   UNDERTALE's parallel story, DELTARUNE. Meet ne...  \
649                                                 NaN   
1252  An expansion to Elden Ring setting players on ...   

                                                Reviews Plays Playing   
644   ['Spamton is so hot, I want to kiss him in the...  1.3K      83  \
649                                                  []     3       0   
1252  ['I really loved that they integrated Family G...     1       0   

     Backlogs Wishlist  
644       468      617  
649       209      644  
1252       39      146

У 3-х игр дата релиза неизвестна. Погуглив выяснил, что Death Stranding 2 и Elden Ring: Shadow of the Erdtree ещё не вышли. А Deltarune вышел на хеллуин 31 октября 2018 года, поэтому невышедшие удалю, а в Deltarune поставлю корректную дату.

In [39]:
df.loc[644,'Release Date'] = 'Oct 31, 2018'
df = df.drop([649,1252]).drop(columns='Unnamed: 0').reset_index(drop=True)
df['Release Date'] = pd.to_datetime(df['Release Date'])

Переведу числа из текстового формата K (тысячи) в целочисленный формат.

In [40]:
k_cols = ['Times Listed',
          'Number of Reviews',
          'Plays', 
          'Playing',
          'Backlogs', 
          'Wishlist']

for col in k_cols:
    df[col] = (df[col].str.replace('K','').astype(float)*1000).astype(int)

In [41]:
df.head(1).T

0
Title                                                     Elden Ring
Release Date                                     2022-02-25 00:00:00
Team                  ['Bandai Namco Entertainment', 'FromSoftware']
Rating                                                           4.5
Times Listed                                                    3900
Number of Reviews                                               3900
Genres                                          ['Adventure', 'RPG']
Summary            Elden Ring is a fantasy, action and open world...
Reviews            ["The first playthrough of elden ring is one o...
Plays                                                          17000
Playing                                                         3800
Backlogs                                                        4600
Wishlist                                                        4800

## Удаление дубликатов

In [42]:
df.duplicated().sum()

382

382 Полных дубликата, так как они не предоставляют ценности, то удалю их.

In [43]:
df = df.drop(columns='Reviews').drop_duplicates().reset_index(drop=True)

## Обработка жанров и команд

In [44]:
df['Team'] = df['Team'].apply(lambda x: ast.literal_eval(x))
team = pd.DataFrame({
    'Title': df['Title'].tolist(),
    'Team': df['Team'].tolist() }).explode('Team').drop_duplicates()

team.to_csv('teams.csv')

team

Title   
0                                  Elden Ring  \
0                                  Elden Ring   
1                                       Hades   
2     The Legend of Zelda: Breath of the Wild   
2     The Legend of Zelda: Breath of the Wild   
...                                       ...   
1114                        Team Sonic Racing   
1115                           Dragon's Dogma   
1116                          Baldur's Gate 3   
1117                 The LEGO Movie Videogame   
1117                 The LEGO Movie Videogame   

                                     Team  
0              Bandai Namco Entertainment  
0                            FromSoftware  
1                        Supergiant Games  
2                                Nintendo  
2     Nintendo EPD Production Group No. 3  
...                                   ...  
1114                                 Sega  
1115                               Capcom  
1116                       Larian Studios  
1117                             WB Games  
1117                            TT Fusion  

[1982 rows x 2 columns]

In [45]:
print('Топ 10 команд по количеству выпущенных игр')
team['Team'].value_counts().head(10)

Топ 10 команд по количеству выпущенных игр


Team
Nintendo                      170
Capcom                         62
Square Enix                    59
Sega                           45
Electronic Arts                42
Konami                         31
Activision                     30
Ubisoft Entertainment          27
Bandai Namco Entertainment     27
Atlus                          24
Name: count, dtype: int64

In [46]:
df['Genres'] = df['Genres'].apply(lambda x: ast.literal_eval(x))

In [47]:
genres = pd.DataFrame({
    'Title': df['Title'].tolist(),
    'Genres': df['Genres'].tolist() }).explode('Genres')

genres

Title               Genres
0                   Elden Ring            Adventure
0                   Elden Ring                  RPG
1                        Hades            Adventure
1                        Hades              Brawler
1                        Hades                Indie
...                        ...                  ...
1116           Baldur's Gate 3             Strategy
1116           Baldur's Gate 3             Tactical
1116           Baldur's Gate 3  Turn Based Strategy
1117  The LEGO Movie Videogame            Adventure
1117  The LEGO Movie Videogame               Puzzle

[2686 rows x 2 columns]

In [48]:
genres = genres.drop_duplicates()
genres['Genres'].value_counts()

Genres
Adventure              711
RPG                    361
Shooter                255
Platform               236
Indie                  207
Puzzle                 129
Strategy               115
Brawler                105
Simulator               92
Turn Based Strategy     70
Fighting                63
Arcade                  61
Visual Novel            52
Tactical                42
Racing                  37
Point-and-Click         36
Music                   21
Sport                   20
Card & Board Game       13
Real Time Strategy       9
Quiz/Trivia              3
MOBA                     2
Pinball                  1
Name: count, dtype: int64

In [49]:
g = genres.copy()
g['Nan'] = np.where(genres['Genres'].isna(),0,1)

for title in  g[g['Nan'] == 0]['Title']:
    display(df[df['Title']==title])

Title Release Date           Team  Rating   
489  Final Fantasy VII Rebirth   2023-12-31  [Square Enix]    -1.0  \

     Times Listed  Number of Reviews Genres   
489        192000             192000     []  \

                                               Summary  Plays  Playing   
489  This next standalone chapter in the FINAL FANT...  20000     3000  \

     Backlogs  Wishlist  
489    354000      1100

Title Release Date                    Team  Rating   
916  Armored Core 2   2000-10-26  [FromSoftware, Agetec]     3.6  \

     Times Listed  Number of Reviews Genres   
916         92000              92000     []  \

                                               Summary   Plays  Playing   
916  The genuine sequel to the original Armored Cor...  278000    18000  \

     Backlogs  Wishlist  
916    232000    119000

Title Release Date            Team  Rating  Times Listed   
1082  Peggle Deluxe   2007-03-06  [PopCap Games]     3.7         69000  \

      Number of Reviews Genres   
1082              69000     []  \

                                                Summary   Plays  Playing   
1082  Take your best shot with energizing arcade fun...  939000    18000  \

      Backlogs  Wishlist  
1082     70000     13000

Есть три пропущенных значения у давольно популярных игр. Заменю их в ручную.

In [50]:
genres.loc[genres['Title'] == 'Final Fantasy VII Rebirth', 'Genres'] = 'RPG'
genres.loc[genres['Title'] == 'Armored Core 2', 'Genres'] = 'Shooter'
genres.loc[genres['Title'] == 'Peggle Deluxe', 'Genres'] = 'Puzzle'

Посомтрим на жанры в которых меньше 10 игр.

In [51]:
rts = genres.query('Genres == "Real Time Strategy"')['Title']
for title in rts:
    display(genres[genres['Title']==title])

Title              Genres
184  Plants vs. Zombies              Puzzle
184  Plants vs. Zombies  Real Time Strategy
184  Plants vs. Zombies           Simulator
184  Plants vs. Zombies            Strategy
184  Plants vs. Zombies            Tactical

Title              Genres
466  Pikmin 2           Adventure
466  Pikmin 2  Real Time Strategy

Title              Genres
530  Pikmin 3           Adventure
530  Pikmin 3  Real Time Strategy

Title              Genres
779  Spore           Adventure
779  Spore  Real Time Strategy
779  Spore                 RPG
779  Spore           Simulator
779  Spore            Strategy

Title              Genres
890  RimWorld               Indie
890  RimWorld  Real Time Strategy
890  RimWorld           Simulator
890  RimWorld            Strategy

Title              Genres
917  Clash of Clans  Real Time Strategy
917  Clash of Clans            Strategy

Title              Genres
988  Warcraft III: The Frozen Throne  Real Time Strategy
988  Warcraft III: The Frozen Throne            Strategy

Title              Genres
996  Plants vs. Zombies 2: It's About Time           Adventure
996  Plants vs. Zombies 2: It's About Time  Real Time Strategy
996  Plants vs. Zombies 2: It's About Time            Strategy
996  Plants vs. Zombies 2: It's About Time            Tactical

Title              Genres
1098  Europa Universalis IV  Real Time Strategy
1098  Europa Universalis IV           Simulator
1098  Europa Universalis IV            Strategy

Судя по всему многие RTS не помечены как Real Time Strategy а остались в категории Strategy, к примеру здесь нет именитого Star Craft (хотя в датасете он есть). Заменю значения на Strategy и удалю дубликаты.

In [52]:
genres['Genres'] = genres['Genres'].replace('Real Time Strategy','Strategy')
genres = genres.drop_duplicates()

C:\Users\elpiz\AppData\Local\Temp\ipykernel_11708\2145279999.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  genres['Genres'] = genres['Genres'].replace('Real Time Strategy','Strategy')


In [53]:
quiz = genres.query('Genres == "Quiz/Trivia"')['Title']
for title in quiz:
    display(genres[genres['Title']==title])

Title       Genres
257  Wordle       Puzzle
257  Wordle  Quiz/Trivia

Title       Genres
951  The Jackbox Party Pack 3        Indie
951  The Jackbox Party Pack 3       Puzzle
951  The Jackbox Party Pack 3  Quiz/Trivia
951  The Jackbox Party Pack 3     Strategy

Title       Genres
1087  The Jackbox Party Pack        Indie
1087  The Jackbox Party Pack       Puzzle
1087  The Jackbox Party Pack  Quiz/Trivia
1087  The Jackbox Party Pack     Strategy

Все квизы являються и пазлами, можно убрать этот жанр с помощью фильтра в DataLens.

In [54]:
moba = genres.query('Genres == "MOBA"')['Title']
for title in moba:
    display(df[df['Title']==title])

Title Release Date                                   Team   
109  League of Legends   2009-10-27  [GOA Games Services Ltd., Riot Games]  \

     Rating  Times Listed  Number of Reviews                 Genres   
109     2.1        484000             484000  [MOBA, RPG, Strategy]  \

                                               Summary  Plays  Playing   
109  League of Legends is a fast-paced, competitive...  13000     1200  \

     Backlogs  Wishlist  
109    324000     52000

Title Release Date                         Team  Rating   
681  Pokémon Unite   2021-07-20  [Timi, The Pokémon Company]     2.7  \

     Times Listed  Number of Reviews  Genres   
681        285000             285000  [MOBA]  \

                                               Summary  Plays  Playing   
681  Team up and take down the opposition in Pokémo...   4500   263000  \

     Backlogs  Wishlist  
681    184000     65000

Всего 2 игры в жанре MOBA, наврятли такое маленькое количество позволит обьективно оценить этот жанр, поэтому этот жанр как и квизы можно отфлитровать.

In [55]:
pinball = genres.query('Genres == "Pinball"')['Title']
for title in pinball:
    display(df[df['Title']==title])

Title Release Date   
1072  Pokémon Pinball: Ruby & Sapphire   2003-08-01  \

                                 Team  Rating  Times Listed   
1072  [Nintendo, Jupiter Corporation]     3.5        135000  \

      Number of Reviews                        Genres   
1072             135000  [Adventure, Arcade, Pinball]  \

                                                Summary  Plays  Playing   
1072  Pokemon Pinball has all the features you'd dem...   1600    23000  \

      Backlogs  Wishlist  
1072    136000     69000

Так же всего одна игра и посути это поджанр аркады, добавлю в фильтр.

Сохраню предобработанный датасет

In [56]:
genres.to_csv('genres.csv')

df.to_csv('games_preprocessed.csv')

## Итоги

- Провёл обзор данных.
- Удалил дубликаты.
- Заполнил пропуски.
- Обработал дату релиза.
- Перевёл численные значения из строкового формата *K (тыс.) в целочисленный формат *000. (К примеру, было 18.5K, стало 18500)
- Так как одна и та же игра может относиться к нескольким жанрам или командам разработчиков, то я выделил ещё два датасета "teams.csv" и "genres.csv".

Данные готовы к визуализации, которую буду делать в Yandex DataLens.